src/abstraction/ak.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    constants::{
+        algorithm::{AsymmetricAlgorithm, Cipher, HashingAlgorithm, SignatureScheme},
+        tss::*,
+        types::session::SessionType,
+    },
+    handles::{AuthHandle, KeyHandle},
+    structures::{Auth, CreateKeyResult, Private},
+    tss2_esys::{
+        TPM2B_PUBLIC, TPMS_ECC_PARMS, TPMS_RSA_PARMS, TPMS_SCHEME_HASH, TPMT_ECC_SCHEME,
+        TPMT_KDF_SCHEME, TPMT_RSA_SCHEME, TPMT_SYM_DEF_OBJECT, TPMU_ASYM_SCHEME, TPMU_SYM_KEY_BITS,
+        TPMU_SYM_MODE,
+    },
+    utils::{
+        ObjectAttributes, PublicIdUnion, PublicParmsUnion, Tpm2BPublicBuilder, TpmaSessionBuilder,
+    },
+    Context, Error, Result, WrapperErrorKind,
+};
+
+fn test_algs_compatibility(key_alg: AsymmetricAlgorithm, sign_alg: SignatureScheme) -> Result<()> {
+    match key_alg {
+        AsymmetricAlgorithm::Rsa => match sign_alg {
+            SignatureScheme::RsaPss | SignatureScheme::RsaSsa => Ok(()),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        },
+        AsymmetricAlgorithm::Ecc => match sign_alg {
+            SignatureScheme::EcDaa
+            | SignatureScheme::EcDsa
+            | SignatureScheme::EcSchnorr
+            | SignatureScheme::Sm2 => Ok(()),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        },
+    }
```

Comment:
I think it might be a better idea to instead add a function like `impl SignatureScheme { fn get_key_alg(&self) -> AsymmetricAlgorithm {... } }`. So basically, ensure you can't make these types of mistakes by just determining the `key_alg` by the `SignatureScheme`. Since it's not common you want to use `EcDsa` with `Rsa` anyway.

Reply:
Also, fun fact: there *are* [`RsaSignatureScheme`](https://github.com/parallaxsecond/rust-tss-esapi/blob/master/src/constants/algorithm.rs#L302) and [`EccSignatureScheme`](https://github.com/parallaxsecond/rust-tss-esapi/blob/master/src/constants/algorithm.rs#L336) enums.

Reply:
> Also, fun fact: there are RsaSignatureScheme and EccSignatureScheme enums.

That's... kinda sad, that we have both. Should trim it down at some point. Which way would you lean in this?

I think I'd be ok with that `fn get_key_alg` and then only asking for a `SignatureScheme` here.

Reply:
I decided to both add `get_key_alg`, and fix the `SignatureScheme` definition.

Reply:
👌🏻 going that extra mile